gusto/linear_solvers.py

```diff
@@ -18,38 +18,34 @@ class TimesteppingSolver(object, metaclass=ABCMeta):
     This is a dummy base class.
 
     :arg state: :class:`.State` object.
-    :arg params (optional): solver parameters
+    :arg solver_params (optional): solver parameters
+    :arg overwrite_solver_parameters: boolean, if True use only the
+    solver_parameters that have been passed in, if False then update
+    the default solver parameters with the solver_parameters passed in.
     """
 
-    def __init__(self, state, params=None):
+    def __init__(self, state, solver_parameters=None,
+                 overwrite_solver_parameters=False):
 
         self.state = state
 
-        if params is None:
-            self.params = {'pc_type': 'fieldsplit',
-                           'pc_fieldsplit_type': 'schur',
-                           'ksp_type': 'gmres',
-                           'ksp_max_it': 100,
-                           'ksp_gmres_restart': 50,
-                           'pc_fieldsplit_schur_fact_type': 'FULL',
-                           'pc_fieldsplit_schur_precondition': 'selfp',
-                           'fieldsplit_0_ksp_type': 'preonly',
-                           'fieldsplit_0_pc_type': 'bjacobi',
-                           'fieldsplit_0_sub_pc_type': 'ilu',
-                           'fieldsplit_1_ksp_type': 'preonly',
-                           'fieldsplit_1_pc_type': 'gamg',
-                           'fieldsplit_1_mg_levels_ksp_type': 'chebyshev',
-                           'fieldsplit_1_mg_levels_ksp_chebyshev_esteig': True,
-                           'fieldsplit_1_mg_levels_ksp_chebyshev_esteig_random': True,
-                           'fieldsplit_1_mg_levels_ksp_max_it': 1,
-                           'fieldsplit_1_mg_levels_pc_type': 'bjacobi',
-                           'fieldsplit_1_mg_levels_sub_pc_type': 'ilu'}
+        if solver_parameters is not None:
+            if overwrite_solver_parameters:
+                self.solver_parameters = solver_parameters
+            else:
+                self.default_solver_parameters.update(solver_parameters)
+                self.solver_parameters = self.default_solver_parameters
```

Comment:
Careful, this doesn't do what you want (since `default_solver_parameters` is a class property).

I think you could do this:

```python
class Solver(...):
     def __init__(self, ..., solver_parameters=None):
        if solver_parameters is not None:
            if not overwrite:
               p = dict(self.solver_parameters)
               p.update(solver_parameters)
               solver_parameters = p
            self.solver_parameters = solver_parameters
            
    @abstractproperty
    def solver_parameters(self):
        """"Solver parameters for this solver"""

class FooSolver(Solver):
    solver_parameters = {...}
```

The lookup order is (simplified)

```
getattr(instance, "solver_parameters")
getattr(instance.__class__, "solver_parameters")
```

So the instance setting overrides the class instance, but if you didn't set something on the instance then it gets looked up on the class.